### DISCUSSION TITLE
How to update Engine credentials inside of Session

### DISCUSSION THREAD

**Discussion by patryk-b-jrdltd** [sqlalchemy/sqlalchemy#7148] [Usage Questions]
I have this issue, where my created event does not update the Engine's password inside of my Session.

I have this file `db.py` which has the code:
```python
__Engine = __connect_to_mysql_db() # this returns a create_engine()
Session = sessionmaker(bind=__Engine)

@event.listens_for(__Engine, 'do_connect')
def provide_token(dialect, conn_rec, cargs, cparams):
    cparams['password'] = get_authentication_token()

# To check whether the password has been updated inside the session
@event.listens_for(Session, 'do_orm_execute')
def receive_do_orm_execute(orm_execute_state):
    print(orm_execute_state.session.bind.url.password)
```
My `create_engine` is like that:
```python
create_engine(engine_url, pool_pre_ping=True, connect_args={'ssl_ca': cert_path}, pool_recycle=5)
```

This is done, because I want to connect to AWS RDS instance, but with each query to have a new password.

The `Session` is imported into the 1db_queries.py1 file and executed as so:
```python
def get_information(item_id: int):
    with Session() as session:
        result = session.query(
            Item
        ).filter(
            Item.id == item_id
        ).first()

    return result
```
How can I update the password of the Engine inside the Session, so that every time my query is executed it has a new password?

**Comment by zzzeek** [comment]
hi there -

that "password" is only passed to the DBAPI when it first creates a connection.  There's no "password" to be sent on "each query" so what you are asking is not possible.

**Reply by CaselIT** [reply]
To clarify, the same session will keep the same connection for all its execution needs, but each time a session is created a new connection is created, and with it a new password can be generated.

Also for this use case it's probably best to use a null pool when creating the engine

**Reply by zzzeek** [reply]
> 
> Also for this use case it's probably best to use a null pool when creating the engine

this is only if the RDS connection has some extremely short timeout, and for that you can just as easily and more efficiently configure [pool_recycle](https://docs.sqlalchemy.org/en/14/core/engines.html?highlight=pool_recycle#sqlalchemy.create_engine.params.pool_recycle)

**Reply by CaselIT** [reply]
By "use case" I meant "use a new password for each execution".

Otherwise I agree pool recycle is best to control the ttl of the connections in the pool

**Reply by patryk-b-jrdltd** [reply]
@CaselIT I don't quite understand. Doesn't the `with Session() as session` create a new session? If not, that means I would have to execute the `sessionmaker` each time?

**Reply by CaselIT** [reply]
>  Doesn't the `with Session() as session` create a new session?

yes, it does. That's not what you asked in the opening:

> How can I update the password of the Engine inside the Session, so that every time my query is executed it has a new password?

that is not supported, since a single session can be used for any number of executions, not only a single query.

To recap:
- a single session instance uses the same connection until it's closed, so it's not possible to change password in it
- each new session ask for a connection before first use, so if needed a new password may be used to establish it via the on_connect method

**Reply by patryk-b-jrdltd** [reply]
> Doesn't the with Session() as session create a new session?
> yes, it does. That's not what you asked in the opening:

Based on what you say, it seems that the password should change inside the `with Session() as session` but printing the password using `print(session.bind.url.password)` returns the same password each time. 

> a single session instance uses the same connection until it's closed

The context manager doesn't close it automatically?

**Reply by CaselIT** [reply]
> Based on what you say, it seems that the password should change inside the `with Session() as session`

indeed, but the `url` on the engine is static, since you can have any number of concurrent session open from the same engine. You should place the print inside the `on_connect` event.

> The context manager doesn't close it automatically?

sure
